Issue #3223: Adding environment variable substitution support for cake.config

epaulsen opened the pull request:
This pullrequests adds basic environment variable substitution to `cake.config`, as requested in #3223 

I've used the windows syntax for environment variables, as that is what nuget does for `nuget.config`.
([See Microsoft docs on this topic](https://docs.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior#environment-variables-in-configuration))
The effect is that environment variable substitution will work cross-platform, but windows environment variable syntax must be used.  This might look somewhat odd in unix environments, but can be justified with the .ini syntax in cake.config 😄 

This should probably be documented somehow, but I can't find any documentation in this repo?

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=3234) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=3234) before we can accept your contribution.<br/><br/>:x: epaulsen [sign now](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=3234)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/cake-build/cake?pullRequest=3234) it.</sub>

augustoproiete commented on src/Cake.Core/Configuration/Parser/ConfigurationParser.cs:
```diff
@@ -37,7 +37,7 @@ public ConfigurationParser(IFileSystem fileSystem, ICakeEnvironment environment)
             using (var reader = new StreamReader(stream))
             {
-                return Read(reader.ReadToEnd());
+                return Read(EnvironmentVariableSubstitution.Substitute(_environment, reader.ReadToEnd()));
```
I took a cursory look at your `EnvironmentVariableSubstitution.Substitute` and it seems that it's re-implementing an existing .NET feature -> [`Environment.ExpandEnvironmentVariables`](https://docs.microsoft.com/en-us/dotnet/api/system.environment.expandenvironmentvariables?view=net-5.0) is that correct?

If that's true, could we use the built-in function instead? Otherwise, what's different/better about this implementation?

epaulsen replied:
Oh.My.Lord.  Well, this is awkward... 
I've been programming .net for the last twenty years, and I wasn't aware that the framework had the functionalty built-in.
Oh well, at least I learned something by doing this 😁 

I'll fix this, I think using built-in functionality would be better.

epaulsen replied:
Rewritten to use built-in functionality instead

augustoproiete commented:
Thanks for the PR @epaulsen!

Re: docs, I think a good place to explain we support environment variable expansion would be in the docs of the website: https://cakebuild.net/docs/running-builds/configuration/set-configuration-values

The site is on a separate repo: https://github.com/cake-build/website

The config docs are here, in markdown: https://github.com/cake-build/website/tree/master/input/docs/running-builds/configuration

augustoproiete commented on src/Cake.Core/Configuration/Parser/ConfigurationParser.cs:
```diff
@@ -37,7 +37,7 @@ public ConfigurationParser(IFileSystem fileSystem, ICakeEnvironment environment)
             using (var reader = new StreamReader(stream))
             {
-                return Read(reader.ReadToEnd());
+                return Read(Environment.ExpandEnvironmentVariables(reader.ReadToEnd()));
```
The scope of the expansion here seems a little too broad, and I'm not 100% sure if it will work for expanding multiple variables on all platforms, considering different line-endings, and the tests don't cover that.

With the current scope, the following scenarios would be possible:

```ini
[Section]
%A%=%B%
```

or even

```ini
[%A%]
%B%=%C%
```

I'm not sure we'd want to support expansion on the whole file rather than on invididual `values` of each field... We probably only want this one scenario to work:

```ini
[Section]
Key=%VALUE%
```

@cake-build/cake-team What do you think?

I think it would be safer to perform the expansion [directly on each value](https://github.com/cake-build/cake/blob/04a98f8e25f4f676f99edb9aef9ac3e724eb809d/src/Cake.Core/Configuration/Parser/ConfigurationParser.cs#L59) as we parse individual entries:

```csharp
case ConfigurationTokenKind.Value:
     var pair = ParseKeyAndValue(tokens, section);
     result[pair.Key] = _environment.ExpandEnvironmentVariables(pair.Value);
     break;
```

The other good thing is that the [`ConfigurationParser` already has access to `ICakeEnvironment`](ExpandEnvironmentVariables), so we can use that instead of calling `Environment` directly.

epaulsen replied:
Ok, that's reasonable and should cover my needs.
Should be fairly easy to implement.

epaulsen replied:
Done, now changed to using environment variable substitution for values only.  
Sorry about the force-push, I am not used to the workflow for pullrequests on Github repositories.

augustoproiete reviewed (changes requested):
Looking good!

epaulsen commented:
@augustoproiete - Done(I think).  
There's a build failure on AppVeyor which seems unrelated to this PR.

Should I squash the commits in to a single commit or leave it as-is?

augustoproiete commented:
Thanks @epaulsen! I think this is it!
If you could squash the commits and rebase on top of `develop` that would help (and might even fix the AppVeyor build).

---

@cake-build/cake-team Could I get another pair of eyes on this one?

epaulsen commented:
> Thanks @epaulsen! I think this is it!
> If you could squash the commits and rebase on top of `develop` that would help (and might even fix the AppVeyor build).

Rebased and squashed into single commit 👍 
But that didn't fix the appveyor build, which seems to be failing on wrong .net core sdk in `global.json`
Fairly certain that I didn't break that, so ignoring it for now.

augustoproiete reviewed (approved):
LGTM!

The AppVeyor build issue is unrelated. @devlead suggested that it's probably caused because of the branch name having only numbers (the branch name is used to generate version numbers and it seems to be generating a version number not considered valid)

augustoproiete commented:
@epaulsen your changes have been merged, thanks for your contribution :+1:
